src/model/operation/batchnorm.cc

```diff
@@ -0,0 +1,177 @@
+#include "./batchnorm.h"
+
+namespace singa {
+
+BatchNormHandle::BatchNormHandle(const float momentum, const Tensor& input, const Tensor& RunningMean,
+                                 const Tensor& RunningVariance) {
+  factor = momentum;
+  batchsize = input.shape(0);
+  channels = input.shape(1);
+  if (input.nDim() == 4u) {
+    height = input.shape().at(2);
+    width = input.shape().at(3);
+    is_2d = false;
+  } else if (input.nDim() == 2u) {
+    height = 1;
+    width = 1;
+    is_2d = true;
+  } else {
+    LOG(FATAL) << "The dimension of input should either be 4D or 2D.";
+  }
+  runningMean = RunningMean;
+  runningVariance = RunningVariance;
+};
+
+CudnnBatchNormHandle::CudnnBatchNormHandle(const float momentum, const Tensor& input, const Tensor& RunningMean,
+    const Tensor& RunningVariance): BatchNormHandle(momentum, input, RunningMean, RunningVariance) {
+  if (is_2d)
+    mode = CUDNN_BATCHNORM_PER_ACTIVATION;
+  else
+    mode = CUDNN_BATCHNORM_SPATIAL;
+  DataType dtype = input.data_type();
+  CUDNN_CHECK(cudnnCreateTensorDescriptor(&shape_desc));
+  CUDNN_CHECK(cudnnCreateTensorDescriptor(&param_desc));
+  CUDNN_CHECK(cudnnSetTensor4dDescriptor(shape_desc, CUDNN_TENSOR_NCHW,
+                                         GetCudnnDataType(dtype),
+                                         batchsize,
+                                         channels, height, width));
+  CUDNN_CHECK(cudnnSetTensor4dDescriptor(param_desc, CUDNN_TENSOR_NCHW,
+                                         GetCudnnDataType(dtype), 1, channels,
+                                         1, 1));
+};
+
+Tensor GpuBatchNormForwardTraining(const Tensor& x, const Tensor& bnScale, const Tensor& bnBias,
+                                   const std::vector<Tensor>& cache, const CudnnBatchNormHandle &cbnh) {
+  CHECK_EQ(x.device()->lang(), kCuda);
+  CHECK_EQ(bnScale.device()->lang(), kCuda);
+  CHECK_EQ(bnBias.device()->lang(), kCuda);
+  CHECK_EQ(cbnh.runningMean.device()->lang(), kCuda);
+  CHECK_EQ(cbnh.runningVariance.device()->lang(), kCuda);
+  CHECK_EQ(cache[1].device()->lang(), kCuda);  //resultmean
+  CHECK_EQ(cache[2].device()->lang(), kCuda);  //resultvariance
+
```

Comment:
what's the difference between runningMean and resultmean?

Reply:
runningMean is a statistics inference of mean of whole dataset used for training. resultMean is the exact mean for the current batch of data. resultMean and resultVariance are needed when do backward, if not stored, they needs to be calculated again in backward(). resultMean is also used to estimate(calculate) runningMean, the equation is runningmean = momentum*runningmean + (1-momentum)* resultmean.